### PR TITLE
plan: collect visit info during logical plan build

### PR DESCRIPTION
**Pull request by tiancaiamao**
Privilege check need DB Table and Column names, and also statement type to decide the privilege type.
In theory, it's possible to get all the information during preprocess, after resolve name. 
But I find many field in the `ast.Node` is still nil.
This is the skeleton code, I'll back to it and add more details step by step.

### CODE REVIEW

File: plan/preprocess.go [hunk 1]
```diff
@@ -18,12 +18,98 @@ import (
...
+		for _, t := range v.Tables {
+			pc.check(t, mysql.DropPriv)
+		}
+	case *ast.DropIndexStmt:
+		pc.check(v.Table, mysql.IndexPriv)
+	case *ast.FieldList:
+	case *ast.InsertStmt:
+	case *ast.LoadDataStmt:
+	case *ast.SelectStmt:
+		for _, rf := range v.GetResultFields() {
```
Comment (hanfei1991) [inline plan/preprocess.go]: Not all used table will appear in select stmt, Like `select a.* from a join b`, You can checkout it in `buildDataSource` in logical plan builder.

**Comment by hanfei1991** [comment]
I think you should check it in logical plan builder phase.

**Comment by tiancaiamao** [comment]
I've add a visitInfo field to logical plan builder, table visit information will be collect during
logical plan builder phase, so privilege checker can use it later.

PTAL @hanfei1991

File: plan/logical_plan_builder.go [hunk 2]
```diff
@@ -1139,6 +1153,15 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 	if b.err != nil {
 		return nil
 	}
+
+	if ds, ok := p.(*DataSource); ok {
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: Does it work ? p will never be a datasource

Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: I'm not sure, but it seems work, see test case `TestVisitInfo`.

File: plan/logical_plan_builder.go [hunk 3]
```diff
@@ -1071,6 +1076,15 @@ func (b *planBuilder) buildUpdate(update *ast.UpdateStmt) LogicalPlan {
...
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: You should check every table refs instead of check datasource.

File: plan/logical_plan_builder.go (hunk 2 above)
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: you should check every `delete tables`, If the delete table is null, you should check every table refs.

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -137,7 +137,7 @@ func (er *expressionRewriter) constructBinaryOpFunction(l expression.Expression,
 func (er *expressionRewriter) buildSubquery(subq *ast.SubqueryExpr) LogicalPlan {
 	outerSchema := er.schema.Clone()
 	er.b.outerSchemas = append(er.b.outerSchemas, outerSchema)
-	np := er.b.buildResultSetNode(subq.Query)
+	np := er.b.buildResultSetNode(subq.Query, 0)
```
Comment (shenli) [inline plan/expression_rewriter.go]: Can we make 0 a const?

File: plan/logical_plan_builder.go [hunk 5]
```diff
@@ -98,19 +98,19 @@ func (b *planBuilder) buildAggregation(p LogicalPlan, aggFuncList []*ast.Aggrega
 	return agg, aggIndexMap
 }
 
-func (b *planBuilder) buildResultSetNode(node ast.ResultSetNode) LogicalPlan {
+func (b *planBuilder) buildResultSetNode(node ast.ResultSetNode, mark mysql.PrivilegeType) LogicalPlan {
```
Comment (shenli) [inline plan/logical_plan_builder.go]: mark is not good name. How about privFlag?

File: plan/logical_plan_builder.go [hunk 6]
```diff
@@ -1133,12 +1147,21 @@ func (b *planBuilder) buildUpdateLists(list []*ast.Assignment, p LogicalPlan) ([
 	return newList, p
 }
 
-func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
+func (b *planBuilder) buildDelete(delete *ast.DeleteStmt, mark mysql.PrivilegeType) LogicalPlan {
 	sel := &ast.SelectStmt{Fields: &ast.FieldList{}, From: delete.TableRefs, Where: delete.Where, OrderBy: delete.Order, Limit: delete.Limit}
-	p := b.buildResultSetNode(sel.From.TableRefs)
+	p := b.buildResultSetNode(sel.From.TableRefs, mark)
```
Comment (shenli) [inline plan/logical_plan_builder.go]: For delete from multiple tables, not all the tables need DeletePriv.

**Comment by shenli** [comment]
Please resolve the conflict.

**Comment by hanfei1991** [comment]
We'd better check ast directly instead of passing a mark arg to every `build****`.
e.g. for `update h1,  (select * from h2) a1 set h1.a = a1.a;`, only h1 need update privilege. Through ast we can know this easily.

File: plan/logical_plan_builder.go (hunk 6 above)
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: I'm not quite understand, could you give a example?

**Comment by tiancaiamao** [comment]
Haven't  handle table name alias in this PR, and more complex test should be added. 
PTAL @hanfei1991

File: plan/logical_plan_builder.go [hunk 7]
```diff
@@ -1156,6 +1174,24 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 	var tables []*ast.TableName
 	if delete.Tables != nil {
 		tables = delete.Tables.Tables
+
+		// Delete a, b from a, b, c, d... add a and b
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: add `.`

File: plan/logical_plan_builder.go [hunk 8]
```diff
@@ -1156,6 +1174,24 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
...
+		for _, table := range tables {
+			b.visitInfo = append(b.visitInfo, visitInfo{
+				privilege: mysql.DeletePriv,
+				db:        table.Schema.L,
+				table:     table.Name.L,
+			})
+		}
+	} else {
+		// Delete from a, b, c, d
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: ditto

File: plan/logical_plan_builder.go [hunk 9]
```diff
@@ -1134,6 +1150,8 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 	if b.err != nil {
 		return nil
 	}
+	tableList := extractTableListFromPlan(p)
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: move this to line 1188

File: plan/logical_plan_builder.go [hunk 10]
```diff
@@ -1168,3 +1204,19 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 	del.SetSchema(expression.NewSchema())
 	return del
 }
+
+func extractTableListFromPlan(p LogicalPlan) []*DataSource {
+	var ret []*DataSource
+	for p != nil {
+		switch v := p.(type) {
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: Why don't you consider Children()[1] ?

File: plan/logical_plan_test.go [hunk 11]
```diff
@@ -1616,3 +1617,151 @@ func (s *testPlanSuite) TestAggPrune(c *C) {
 		c.Assert(ToString(p), Equals, ca.best, comment)
 	}
 }
+
+func (s *testPlanSuite) TestVisitInfo(c *C) {
```
Comment (hanfei1991) [inline plan/logical_plan_test.go]: add test for multi delete/update

File: plan/logical_plan_builder.go (hunk 9 above)
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: p at line 1188 is not the Plan generate from `from xxx`, I need the table list, not consider where,orderby, grou

File: plan/logical_plan_builder.go (hunk 10 above)
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: I guess `delete from a,b,c` generate a list, chained by join's left child , is that true?

File: plan/logical_plan_test.go (hunk 11 above)
Comment (tiancaiamao) [inline plan/logical_plan_test.go]: Would you give some examples, in detail ?

Comment (shenli) [inline plan/logical_plan_test.go]: DELETE FROM a1 USING t1 AS a1 INNER JOIN t2 AS a2 WHERE a1.id=a2.id;

File: plan/logical_plan_builder.go (hunk 10 above)
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: I join tree is like ((((a, b), c), d), e), Only a is left child, b,c,d,e are all right child

File: plan/logical_plan_builder.go (hunk 9 above)
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: I think you'd better extract table info from ast tree.

File: plan/logical_plan_builder.go [hunk 12]
```diff
@@ -1180,3 +1216,19 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
...
+		case *Join:
+			p = v.Children()[0].(LogicalPlan)
+		case *DataSource:
+			ret = append(ret, v)
+			return ret
```
Comment (coocood) [inline plan/logical_plan_builder.go]: So this function return just one *DataSource, then why use a slice?

**Comment by shenli** [comment]
@tiancaiamao Please resolve the conflict.

**Comment by tiancaiamao** [comment]
PTAL @hanfei1991

File: plan/logical_plan_builder.go [hunk 13]
```diff
@@ -942,6 +942,13 @@ func (b *planBuilder) buildDataSource(tn *ast.TableName) LogicalPlan {
 	}
 	p.self = p
 	p.initIDAndContext(b.ctx)
+
+	b.visitInfo = append(b.visitInfo, visitInfo{
```
Comment (zimulala) [inline plan/logical_plan_builder.go]: Can this be extracted into a function? This is used in many places.

File: plan/logical_plan_builder.go [hunk 14]
```diff
@@ -1186,3 +1236,17 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
...
 }
+
+func extractTableList(node ast.ResultSetNode, input []*ast.TableName) []*ast.TableName {
+	switch x := node.(type) {
+	case *ast.Join:
+		input = extractTableList(x.Left, input)
+		input = extractTableList(x.Right, input)
+	case *ast.TableSource:
+		if s, ok := x.Source.(*ast.TableName); ok {
+			fmt.Println("xsdfasdf", s.Name.L)
```
Comment (zimulala) [inline plan/logical_plan_builder.go]: Remove this useless log.

Comment (coocood) [inline plan/logical_plan_builder.go]: Remove this line

File: plan/optimizer.go [hunk 15]
```diff
@@ -66,6 +66,7 @@ func Optimize(ctx context.Context, node ast.Node, is infoschema.InfoSchema) (Pla
 		is:        is,
 		colMapper: make(map[*ast.ColumnNameExpr]int),
 		allocator: allocator,
+		visitInfo: make([]visitInfo, 0, 3),
```
Comment (zimulala) [inline plan/optimizer.go]: Why here is 3? Can it be changed to a constant?  Then add a comment.

File: plan/logical_plan_builder.go [hunk 16]
```diff
@@ -1171,10 +1194,37 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 			return nil
 		}
 	}
+
```
Comment (coocood) [inline plan/logical_plan_builder.go]: remove empty line

File: plan/logical_plan_builder.go [hunk 17]
```diff
@@ -1152,6 +1174,7 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 	if b.err != nil {
 		return nil
 	}
+
```
Comment (coocood) [inline plan/logical_plan_builder.go]: remove empty line

File: plan/preprocess.go [hunk 18]
```diff
@@ -25,5 +25,6 @@ func Preprocess(node ast.Node, info infoschema.InfoSchema, ctx context.Context)
 	if err := ResolveName(node, info, ctx); err != nil {
 		return errors.Trace(err)
 	}
+
```
Comment (coocood) [inline plan/preprocess.go]: remove empty line

File: plan/logical_plan_test.go [hunk 19]
```diff
@@ -1362,3 +1363,165 @@ func (s *testPlanSuite) TestAggPrune(c *C) {
...
+
+func (v visitInfoArray) Swap(i, j int) {
+	v[i], v[j] = v[j], v[i]
+}
+
+func unique(v []visitInfo) []visitInfo {
+	repeat := 0
+	for i := 1; i < len(v); i++ {
+		if v[i] == v[i-1] {
+			repeat++
```
Comment (zimulala) [inline plan/logical_plan_test.go]: There are something wrong with the function of `unique`. I think here use `v = append(v[:i-1], v[i:] ...)` is enough.

File: plan/logical_plan_test.go [hunk 20]
```diff
@@ -1362,3 +1363,165 @@ func (s *testPlanSuite) TestAggPrune(c *C) {
...
+		},
+		{
+			sql: "delete from t where a = 1",
+			ans: []visitInfo{
+				{mysql.DeletePriv, "test", "t", ""},
+				{mysql.SelectPriv, "test", "t", ""},
+			},
+		},
+		{
+			sql: "delete from a1 using t as a1 inner join t as a2 where a1.a = a2.a",
```
Comment (hanfei1991) [inline plan/logical_plan_test.go]: you should add a multi join case like 'delete from t1, t2, t3, t4'. I think you should use different tables.

File: plan/logical_plan_builder.go (hunk 17 above)
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: This new line is necessary.

File: plan/logical_plan_builder.go (hunk 16 above)
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: The new line can split the logic block to make the code more readable.

File: plan/logical_plan_builder.go [hunk 21]
```diff
@@ -1171,10 +1186,27 @@ func (b *planBuilder) buildDelete(delete *ast.DeleteStmt) LogicalPlan {
 			return nil
 		}
 	}
+
 	var tables []*ast.TableName
 	if delete.Tables != nil {
 		tables = delete.Tables.Tables
+		// Delete a, b from a, b, c, d... add a and b.
+		for _, table := range delete.Tables.Tables {
```
Comment (zimulala) [inline plan/logical_plan_builder.go]: s/tables/delete.Tables.Tables

File: plan/logical_plan_test.go (hunk 19 above)
Comment (zimulala) [inline plan/logical_plan_test.go]: This `unique` is used after `sort`.
I got it.

File: plan/logical_plan_test.go (hunk 20 above)
Comment (tiancaiamao) [inline plan/logical_plan_test.go]: The test suite use a mocked infoschema, it add too much complexity to use different tables.

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

File: plan/logical_plan_test.go (hunk 20 above)
Comment (hanfei1991) [inline plan/logical_plan_test.go]: But you still have a bug when processing multi join....

**Pull request merged**